docs/pallas/gpu/pipelining.md

```diff
@@ -0,0 +1,328 @@
+
+- `plgpu.GPUMemroySpace.TMEM` allocates a Ref in Tensor Memory (TMEM). This is a TensorCore Gen 5 (Blackwell) feature for storing arguments and results of asynchronous matrix multiplication instructions. Like GMEM, TMEM cannot be accessed directly but must first be copied to/from SMEM.
+
+The primary purpose of `emit_pipeline` is used to overlap TensorCore computation with data transfers between GMEM and SMEM, since asynchronous copies between GMEM/SMEM have a long latency, but all TensorCore computation must happen in registers or SMEM Refs.
+<!-- #endregion -->
+
+<!-- #region id="0uzcrDCtKABQ" -->
+### Example: Matmul Kernel on Hopper GPUs
+<!-- #endregion -->
+
+<!-- #region id="vILVdlqEdoEK" -->
+Let's begin with a matrix multiplication example designed to run on Hopper GPUs. This kernel utilizes the Hopper-specific `wgmma` (warpgroup matrix multiply accumulate) instruction, which issues a matrix multiplication that runs asynchronously on the TensorCore. Under the hood, `wgmma` is a warpgroup-level instruction in which all 128 CUDA threads in a warpgroup collectively issue a matrix multiplication command. But since a program "thread" in Mosaic GPU conveniently corresponds to a warpgroup, all details of `wgmma` at the CUDA thread level are abstracted away.
+
+This kernel implements a blockwise matrix multiplication of two matrices of shape `[M, K] @ [K, N] = [M, N]`, where each output block is computed in parallel over the CUDA grid. This grid is specified as the `grid` argument to the outer `pl.pallas_call`, and iterates over the non-contracting dimensions M, N of the matrix multiplication.
+<!-- #endregion -->
+
+<!-- #region id="KSvqVNdy726B" -->
+
+![pipeline_matmul](../../_static/pallas/gpu/pipeline_matmul.svg)
+
+<!-- #endregion -->
+
+<!-- #region id="10ebHCQ571Fn" -->
+
+Within a block, we run a sequential pipeline using `plgpu.emit_pipeline` that reduces over the contracting dimension K of the matrix multiplication. On each iteration of the pipeline, we load one tile from each input matrix, multiply them, and then store the result in an accumulator Ref (`plgpu.ACC`). `plgpu.ACC` is a special type of Ref that lives in registers and holds the intermediate results of WGMMA. Once we have accumulated over the entire contracting dimension, we write out the result to the output Ref.
+
+To perform the actual matrix multiplication, we call `plgpu.wgmma` with the accumulator, LHS, and RHS sides as arguments in order to push the arguments into the TensorCore pipeline. Since `wgmma` is an asynchronous instruction, `plgpu.wgmma_wait(N)` is used to wait until there are no more than N `wgmma` operations left in-flight. In this particular implementation we wait for 1 in-flight WGMMA, meaning that the WGMMA we queue on the current iteration will be waited for on the next iteration.
```

Comment:
Should we say that WGMMAs are executed in order? You do say "queue" here, which sort of implies that, so maybe the current is fine too.

Reply:
Done.